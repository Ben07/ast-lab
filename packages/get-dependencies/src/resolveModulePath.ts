import enhancedResolve from 'enhanced-resolve';
import path from 'path';

import isCore from './isCore';
import isThirdParty from './isThirdParty';
import { Alias } from './types';

type ResolveOptions = {
  alias?: Alias,
  moduleDirectory?: string[],
  extensions?: string[],
  plugins?: any[]
}

export default async function resolveModulePath(
  mod: string, 
  source: string, 
  { 
    alias, 
    moduleDirectory: modules, 
    extensions,
    plugins
  }: ResolveOptions = {}
): Promise<string | void> {
  if (!mod || isCore(mod)) {
    return;
  }
  if (path.isAbsolute(mod)) {
    return mod;
  }
  const resolver = enhancedResolve.create({
    extensions: extensions,
    modules,
    alias,
    plugins
  });
  const basedir = path.dirname(source);
  const result = await new Promise((resolve, reject) => resolver(basedir, mod, (err: Error, res: string | undefined) => {
    if (err) {
      return reject(err.message);
    }
    if (!res) {
      console.log(`Couldn't find module ${mod}!`);
      return resolve();
    }
    if (isThirdParty(res)) {
      return resolve();
    }
    return resolve(res);
  })) as string | void;
  return result;
}
