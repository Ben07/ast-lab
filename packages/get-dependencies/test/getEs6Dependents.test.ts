import path from 'path';
import fs from 'fs';
// @ts-ignore
import { enumerateArgsTestFunction, configArgs } from 'lazy-jest';
import _getEs6Dependents from '../src/es6Detect/getEs6Dependents';
import resolveModulePath from '../src/createResolver';

const FIXTURE_DIR = 'test/__fixtures__';
const getPath = (file: string) => path.resolve(FIXTURE_DIR, file);
function getEs6Dependents(file: string, opts: any) {
  return _getEs6Dependents(getPath(file), opts);
}

function loader(dir: string) {
  return async (mod: string) => {
    const realpath = await resolveModulePath()(mod, getPath(`${dir}/index`));
    return realpath ? fs.readFileSync(realpath, 'utf-8') : '';
  };
}
enumerateArgsTestFunction(
  getEs6Dependents,
  configArgs()
  .arg('source', [
    'imports/importNamed.js',
    'imports/importAlias.js',
    'imports/importAll.js',
    'imports/importDefault.js',
    'imports/importDeeperPath.js',
    'imports/noImports.js',
  ])
  .arg('opts', [
    { inDetail: false, loader: loader('imports') },
    { inDetail: true,  loader: loader('imports') }
  ])
);

enumerateArgsTestFunction(
  getEs6Dependents,
  configArgs()
  .arg('source', [
    'exports/exportAllFrom.js',
    'exports/exportDefaultAliasFrom.js',
    'exports/exportNamedFrom.js',
    'exports/noExports.js'
  ])
  .arg('opts', [
    { inDetail: false,  loader: loader('exports') },
    { inDetail: true,  loader: loader('exports') }
  ])
);
