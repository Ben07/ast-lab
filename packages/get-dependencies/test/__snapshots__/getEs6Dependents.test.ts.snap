// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/exportAllFrom.js",{"inDetail":false}] 1`] = `
Map {
  "./exportNamed" => Map {
    "a" => Object {
      "affectedExports": Set {},
      "alias": null,
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/exportAllFrom.js",{"inDetail":true}] 1`] = `
Map {
  "./exportNamed" => Map {
    "a" => Object {
      "affectedExports": Set {},
      "alias": null,
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/exportDefaultAliasFrom.js",{"inDetail":false}] 1`] = `
Map {
  "./exportDefault" => Map {
    "default" => Object {
      "affectedExports": Set {
        "b",
      },
      "alias": "b",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/exportDefaultAliasFrom.js",{"inDetail":true}] 1`] = `
Map {
  "./exportDefault" => Map {
    "default" => Object {
      "affectedExports": Set {
        "b",
      },
      "alias": "b",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/exportNamedFrom.js",{"inDetail":false}] 1`] = `
Map {
  "./exportNamed" => Map {
    "a" => Object {
      "affectedExports": Set {
        "a",
      },
      "alias": "a",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/exportNamedFrom.js",{"inDetail":true}] 1`] = `
Map {
  "./exportNamed" => Map {
    "a" => Object {
      "affectedExports": Set {
        "a",
      },
      "alias": "a",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/noExports.js",{"inDetail":false}] 1`] = `Map {}`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["exports/noExports.js",{"inDetail":true}] 1`] = `Map {}`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importAlias.js",{"inDetail":false}] 1`] = `
Map {
  "./base" => Map {
    "a" => Object {
      "affectedExports": true,
      "alias": "b",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importAlias.js",{"inDetail":true}] 1`] = `
Map {
  "./base" => Map {
    "a" => Object {
      "affectedExports": null,
      "alias": "b",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importAll.js",{"inDetail":false}] 1`] = `
Map {
  "./base" => Map {
    "*" => Object {
      "affectedExports": true,
      "alias": "base",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importAll.js",{"inDetail":true}] 1`] = `
Map {
  "./base" => Map {
    "*" => Object {
      "affectedExports": null,
      "alias": "base",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importDefault.js",{"inDetail":false}] 1`] = `
Map {
  "./base" => Map {
    "default" => Object {
      "affectedExports": true,
      "alias": "base",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importDefault.js",{"inDetail":true}] 1`] = `
Map {
  "./base" => Map {
    "default" => Object {
      "affectedExports": null,
      "alias": "base",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importNamed.js",{"inDetail":false}] 1`] = `
Map {
  "./base" => Map {
    "a" => Object {
      "affectedExports": true,
      "alias": "a",
    },
  },
}
`;

exports[`getEs6Dependents() - (["source", "opts"]) - ["imports/importNamed.js",{"inDetail":true}] 1`] = `
Map {
  "./base" => Map {
    "a" => Object {
      "affectedExports": null,
      "alias": "a",
    },
  },
}
`;
